### ISSUE TITLE
URL with relative uri

### ISSUE THREAD

**Issue by XadillaX** [wintercg/fetch#8]
E.g. `fetch('/foo')`, what should the whole URL should be?

**Comment by lucacasonato** [comment]
We need documentation for what is done by:

- Deno
- Undici
- CFW
- Browsers (main thread)
- Browsers (workers)
- Browsers (service workers)

**Comment by jimmywarting** [comment]
If i could go back and change how node-fetch was first written then i would have made it into a class that you have to construct with some options...

```js
const fetch = new Fetch({
  base: 'http://localhost:8080',
  userAgent,
  cacheStorage,
  ...etc
})

fetch('/foo') // http://localhost:8080/foo
```

**Comment by XadillaX** [comment]
Maybe we can define the `location` not only HTTP or file protocol, but use something like `winter://` or something else.

**Comment by KhafraDev** [comment]
undici will throw an error on relative urls. Browser side it makes sense, but I don't personally think it's a good idea elsewhere.

**Comment by XadillaX** [comment]
> undici will throw an error on relative urls. Browser side it makes sense, but I don't personally think it's a good idea elsewhere.

Think of one situation of Winter's SSR:

If the frontend code like this: (and if the JavaScript is http://foo/index.js)

```js
fetch('/api');
```

it means to `fetch` the data from `http://foo/api`. If Winter support this directly, the frontend JavaScript needn't do any modification when it runs on Winter's SSR. But if it throws error, the developer may have to modify this frontend JavaScript code to adapt Winter's SSR.

**Comment by blittle** [comment]
I plan on documented how all of the platforms behave, but Deno for example, provides a flag for running a script to define the location and subsequently origin. Doing so allows them to handle relative paths. If `--location` is not provided, it throws:

![image](https://user-images.githubusercontent.com/1566869/180321532-eb9db82d-c6d5-4cfb-9129-ebffa41335bb.png)

**Comment by benjamingr** [comment]
@jimmywarting this is the API Node.js core wanted to do initially when we were talking about fetch in 2018 but it's not the API users have been asking for.

**Comment by jimmywarting** [comment]
yea... i know, ppl want something close to the spec as possible....
there could exist an existing (default) `globalThis.fetch` that dose not handle relative urls and also a `globalThis.Fetch` constructor 🤷‍♂️

didn't actually follow that discussion in 2018...
would it be possible to actually have an origin (even if it dose not make sense in the backend?

Kinda like: `node --origin=localhost:8080`
Maybe could be possible to define it inside of package.json too `{ "origin": "localhost:8080" }`?

but perhaps somebody want something more dynamic and programmable and not so static.

---

Something far fetch and possible way for something to actually have an origin would be if someone where to do a http import:
```js
#!/usr/bin/node

import mod from 'https://example.com/module.js'
import github from 'https://github.com/api.js'
```
then `mod` would have a `import.meta.url` equal to: `https://example.com/module.js` 
and github would have `import.meta.url` equal to: `https://github.com/api.js` 

if `github.com/api.js` would call `fetch('/users')` then it could resolve the origin relative to `import.meta.url`
```js
console.assert(import.meta.url === 'https://github.com/api.js')
fetch('/users')  // would fetch https://github.com/users (based on import.meta.url)
```
```js
console.assert(import.meta.url === 'https://example.com/module.js')
fetch('/weather')  // would fetch https://example.com/weather (based on import.meta.url)
```

I have no idea if it would be even at all possible for fetch to resolve something based on `import.meta.url`, we would have to use stack traces to figure out what file/origin called what...

it would certainly be abnormal and not according to spec... and also possible breaking in some cases.
it could certainly be a grate feature but it's not so backward compatible friendly solution either.

maybe something that would be best would be to have an origin on the `Agent` instead

**Comment by benjamingr** [comment]
> yea... i know, ppl want something close to the spec as possible....

No... honestly people didn't care a lot about it being spec compliant that was mostly driven by core members caring. People just didn't want the extra `makeFetch` step and wanted a single function.

---

The "based on module URL if any" suggestion is interesting but I agree it diverges much from what browsers do and I'd rather have no supported behavior than different supported behavior here for server runtimes, probably.

**Comment by Ethan-Arrowood** [comment]
> honestly people didn't care a lot about it being spec compliant that was mostly driven by core members caring.

fwiw we cared about spec compliance because of what we would hear from community. in my experience, its a double edged sword. When we build something spec compliant, users that expect a fetch-like client will complain the loudest. When we build something that is fetch-like, users that expect a compliant client will complain the loudest. We can't really win one way or another (but that is where WinterCG Fetch will change things. we can be fetch-like and spec compliant at the same time!).

So for the purposes of this issue we should remember we are building for runtimes / server environments here.

**Comment by KhafraDev** [comment]
undici supports fetching relative urls now, albeit in a kinda contrived way - 

```js
import { setGlobalOrigin, fetch } from 'undici'

setGlobalOrigin('http://localhost:3000')

const response = await fetch('/api/ping')

console.log(response.url) // http://localhost:3000/api/ping
```
